### ISSUE TITLE
RULE A BLOCKER: tinyllama Q4_K_M produces GPU gibberish with rc=0, http 200, no fallback — CPU passes the same prompt

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#3899]
## Measured at HEAD, not from a stale receipt

`evidence/dogfood/models/0.69.1/lambda.json`, **`apr_sha 712f6d433`** (= HEAD), fresh run:

```
inv:tinyllama-1.1b-chat-v1.0.Q4_K_M.gguf
  green:            false
  qa_rc:            5
  gates_failed:     ['golden_output']
  capability_match: PASSED — "Architecture 'llama': all 4 required ops supported by GPU"
  golden_output:    FAILED — "GPU output failed (CPU passed):
                              golden_output_gpu: gibberish (fragment \"\n[S]\" repeats 3+ times)"
  backends.cuda:    ran=true  fallback=false  rc=0
                    verbs: run rc=0 · chat rc=0 · code rc=0 · serve probed
                    routes: /api/chat both legs http 200 · /v1/chat/completions both legs http 200
```

## Why this is the worst available shape

**Every success signal fires and the output is garbage.**

- `capability_match` **passes** — it affirmatively states all 4 required llama ops are GPU-supported. The PMAT-785 refusal machinery that exists precisely to prevent *"silent Q4_K-decode garbage"* does not trigger, because the gibberish is downstream of capability.
- `fallback: false` — it really ran on the GPU. This is not a mislabelled CPU run.
- `rc=0` on `run`, `chat` and `code`.
- **http 200 on every serve route, both stream legs.**
- CPU passes the **same prompt** — so this is a backend differential, not a prompt artifact.

A user gets a 200 and fluent-looking garbage. Nothing in the stack says otherwise. This is the exact failure mode Rule A exists to forbid:

> **"ALL Q4_K models must work on CUDA on gx10 AND lambda before any release"** — operator, 2026-09-21, *"the end."*

## The differential that narrows it

`tinyllama` is `llama` architecture. **Every other Q4_K model in the fleet inventory is qwen2 / qwen3 / qwen35 — and all 16 pass on CUDA.**

lambda holds it; **gx10 does not** (its 16 Q4_K files are all qwen*). So the fleet has exactly one `llama`-arch Q4_K model and it is the one producing GPU garbage.

**Hypothesis, not a conclusion:** the Q4_K GPU path may be correct for the qwen families and wrong for `llama` — an architecture-conditional defect that 16 passing qwen rows cannot detect. Plausible carriers: head/rope geometry, a tied-vs-untied `lm_head`, or a norm placement that differs between llama and qwen2.

**Verify the differential before adopting it.** Three two-branch decision rules were refuted on their shared premise during this same batch. In particular, re-confirm from the harness that the CPU and GPU legs used the same prompt and the same tokenizer path rather than trusting the message string `"(CPU passed)"`.

Suggested first measurements, cheapest first:

1. Re-run both legs on the same prompt with `--trace` and diff per-layer stats — CPU vs GPU, same file. `"\n[S]" repeating` is *structured* degeneracy (a real token repeating), which points at wrong logits rather than at garbage weights.
2. Vary the prompt (≥4 neighbours). One failing input is an anecdote; that rule already inverted a GPU-correctness diagnosis once in this batch.
3. If the GPU is wrong, bisect where: embedding → per-layer → `lm_head`. A logit-only divergence points at the head; growing divergence points at the backend.

## Second defect visible in the same row

`serve` recorded `teardown: "escalated"` and **`used_gpu: null` on all six routes.** So the route results cannot say whether the GPU served them. A model whose GPU output is garbage, probed over routes that do not record which backend answered, is two gaps stacked. Track separately if it is not the same cause — a `used_gpu: null` is a route that proved nothing about the backend.

## Blocking status

This blocks the 0.69.1 tag. The release judge refuses the row (inventory loop: `check_model_ladder.sh:4-6`, *"every model the host HOLDS"*), **and** it is Rule A.

It must not be resolved with an `inventory.deferred` entry. The operator's ruling on that mechanism covers **MoE** ("we kick the MoE work to later", 2026-09-22); this is a dense `llama` Q4_K model that the build claims full GPU capability for. **Deferring a silent-wrong-answer on a Q4_K model is the precise thing Rule A's "the end." forecloses.**

## Anti-vacuity control (required)

The gate that let this reach a receipt at all must also be fixed, or the next instance is invisible:

1. A row with `capability_match` pass, `fallback: false`, `rc: 0`, all routes 200, and a **failing `golden_output_gpu`** must refuse — and must name the CPU/GPU differential specifically, not just "golden_output failed".
2. `used_gpu: null` must not count as a passing route. A route that cannot say which backend answered is unprobed, not green.
3. Mutant: make the GPU leg pass while CPU fails, and the same guard must fire — the check is *disagreement*, not *GPU-is-bad*.

Related: #3898 (the judge ignores `qa_rc`/`gates_failed`, so this row's `qa_rc: 5` is currently unread by the gate), #3897.

**Comment by noahgift** [comment]
## Closing: NOT a CUDA correctness defect. Rule A is not violated. And this ticket should not have been filed the way it was.

Two independent measurements agree, and #3873 had already reached the conclusion before this was filed.

### Measurement 1 — on the gate's own prompt, there is no differential at all (#3873)

```
apr run (GPU vs CPU)   IDENTICAL output on the same prompt
apr qa with NO CUDA DEVICE VISIBLE (so the CPU check must run):
  golden_output: passed=false
  '<s>[INST] What is the capital of France? [/INST]\n\n[S][INST] France is the capital of France...'
```

**Both backends produce the same degenerate loop. Neither says "Paris".** And the `"GPU output failed (CPU passed)"` clause this ticket quoted as evidence is a **hardcoded format string** in `GpuGoldenLeg::failure()` — nothing checks the CPU leg, and the control flow returns on GPU failure *before* the CPU pattern check ever runs. **Asserted, not measured.**

### Measurement 2 — on a coherent prompt, the one divergence is a near-tie

CPU logits recomputed at exactly the prefix both legs had consumed:

```
d=143   cpu_chose=2982   gpu_chose=26014
logit(2982)  = 14.78509        logit(26014) = 15.02783
gap = 0.24275     top1 = 15.02783, top10 = 11.76790, spread = 3.25994
gap as fraction of local spread = 7.4%
cpu_tok_is_argmax = FALSE      rank_of_gpu_tok = 1
```

0.243 logits against a 3.26 spread. Well inside fp accumulation order. **And the GPU's token is the argmax of the CPU's own recomputed logits** — so if either leg deviates at 143 it is the CPU's incremental path, not the GPU. (Caveat kept from the measurer: `forward_hidden_states` is a full forward and `generate_with_cache` is incremental, so that second point is not a controlled comparison. The 0.24-of-3.26 gap does not depend on resolving it.)

Also established along the way, refuting this ticket's central hypothesis: **`TinyLLama-5M / llama / F16` does not diverge at all** over the same 192-token budget, while `Qwen3-1.7B` and `qwen2.5-1.5B` also do not. A llama-architecture model that does not diverge kills *"the Q4_K GPU path is right for qwen geometry and wrong for llama"*. And the divergent streams **reconverge — 45 of 48 tokens after the split are identical**, which no systematic kernel error produces.

### What I got wrong, and it is a process failure rather than a measurement one

**This ticket's headline evidence was a string that #3873 had already shown to be fabricated.** I even wrote the caution into the ticket — *"re-confirm from the harness that the CPU and GPU legs used the same prompt … rather than trusting the message string `(CPU passed)`"* — and then filed it as a **RULE A BLOCKER**, escalated it as the board's top priority, and redirected a session off other work onto it.

#3873 names the consequence precisely: *"Three sessions were redirected onto it and a release was declared blocked."* That was me.

**Flagging that a claim needs verification is not verifying it.** Writing the caveat and escalating anyway is worse than not noticing, because the caveat makes the escalation look considered. The correct action was to read the open ticket on the same row before filing a second one.

And the direction of the error is the point #3873 makes: a gate that misses a defect costs you the defect; **this one invented the most alarming class available** — silent GPU corruption on the operator's hardest rule — on a model that fails everywhere.

### Disposition, pre-committed and honored

The tinyllama row **does not go silently green and is not deferred.** It splits into two real pieces:

1. **#3873** — `GpuGoldenLeg::failure()` must stop asserting the other leg's verdict, and the `?`-propagation past a check it then reports on must be fixed. Its checklist includes the mandatory red in both directions, which is what preserves the #3477 signal for a case where the CPU genuinely does pass.
2. **A new ticket for why tinyllama degenerates** — Llama-2 `[INST]` templating (it is the only non-qwen model in the inventory and echoes the template back verbatim) and/or special-token decode (`[S]` looks like `<s>`/`</s>` losing its angle brackets). **That is what would make the row green**, and it is a text-generation defect, not a CUDA one.

Superseded by #3873. Refs #3873, #3477, #3900.
